### PR TITLE
gcolumn: add `tidb_enable_unsafe_substitute` to control generated column substitute

### PR DESCRIPTION
**Pull request by xiongjiwei**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v6.3 (TiDB 6.3 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: system-variables.md [hunk 1]
```diff
@@ -2410,3 +2410,10 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
 - 是否持久化到集群：是
 - 默认值：`ON`
 - 这个变量用于控制计算窗口函数时是否采用高精度模式。
+
+### `tidb_enable_unsafe_substitute`
```
Comment (shichun-0415) [inline system-variables.md:2414]: Please specify the introduction version of this variable.

File: system-variables.md [hunk 2]
```diff
@@ -2410,3 +2410,10 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`OFF`
+- 这个变量用于控制是否对生成列中表达式替换使用不安全的替换方式。
```
Comment (wjhuang2016) [inline system-variables.md:2419]: 加个链接解释说明？

File: generated-columns.md [hunk 3]
```diff
@@ -149,6 +149,8 @@ desc select a+1 from t where a+1=3;
 > 上例中，`a` 的类型是 int，而 `a+1` 的列类型是 bigint，如果将生成列的类型定为 int，就不会发生替换。
 >
 > 关于类型转换规则，可以参见[表达式求值的类型转换](/functions-and-operators/type-conversion-in-expression-evaluation.md)。
+> 
+> 若待替换的表达式类型和生成列类型都是字符类型，且只在长度不同时可使用系统变量 `tidb_enable_unsafe_substitute` 来允许其替换。在使用此系统变量时用户需要自己保证生成列计算得到的值严格满足生成列的定义，否则有可能得到错误的[结果](https://github.com/pingcap/tidb/issues/35490#issuecomment-1211658886)。
```
Comment (wjhuang2016) [inline generated-columns.md:153]: 详细说下什么情况下会出问题

File: system-variables.md [hunk 4]
```diff
@@ -2410,3 +2410,10 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
 - 是否持久化到集群：是
 - 默认值：`ON`
 - 这个变量用于控制计算窗口函数时是否采用高精度模式。
+
+### `tidb_enable_unsafe_substitute` <span class="version-mark">从 v6.3.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`OFF`
+- 这个变量用于控制是否对生成列中表达式替换使用不安全的替换方式。详情见[生成列](/generated-columns.md)。
```
Comment (shichun-0415) [inline system-variables.md:2419]: `unsafe` 和 `不安全` 这种字眼感觉会传达负面信息，可以改个名字不呢？
```suggestion
- 这个变量用于控制是否对生成列中表达式替换使用不安全的替换方式。默认值为 `OFF`，即默认关闭不安全的替换方式。详情见[生成列](/generated-columns.md)。
```

Comment (xiongjiwei) [reply system-variables.md:2419]: 这个变量确实是不安全的，有可能导致查询结果出错

**Review by shichun-0415 — APPROVED** [review]
LGTM

File: generated-columns.md [hunk 5]
```diff
@@ -111,15 +111,18 @@ ERROR 1048 (23000): Column 'city' cannot be null
 
 ## 索引生成列替换
 
-当查询中出现的某个表达式与一个含索引的生成列同等时，TiDB 会将这个表达式替换为对应的生成列，这样就可以在生成查询计划时考虑使用这个索引。
+当查询中出现的某个表达式与一个含索引的生成列严格同等时，TiDB 会将这个表达式替换为对应的生成列，这样就可以在生成查询计划时考虑使用这个索引。
 
-例如，下面的例子为 `a+1` 这个表达式创建生成列并添加索引，从而加速了查询。
+下面的例子为 `a+1` 这个表达式创建生成列并添加索引，从而加速了查询。其中，`a` 的类型是 int，而 `a+1` 的列类型是 bigint，如果将生成列的类型定为 int，就不会发生替换。关于类型转换规则，可以参见[表达式求值的类型转换](/functions-and-operators/type-conversion-in-expression-evaluation.md)。
```
Comment (TomShawn) [inline generated-columns.md:116]:
```suggestion
下面的例子为 `a+1` 这个表达式创建生成列并添加索引，从而加速查询。其中，`a` 的列类型是 int，而 `a+1` 的列类型是 bigint。如果将生成列的类型改为 int，就不会发生替换。关于类型转换的规则，可以参见[表达式求值的类型转换](/functions-and-operators/type-conversion-in-expression-evaluation.md)。
```

File: generated-columns.md [hunk 6]
```diff
@@ -144,11 +152,7 @@ desc select a+1 from t where a+1=3;
 
 > **注意：**
 >
-> 只有当待替换的表达式类型和生成列类型严格相等时，才会进行转换。
->
-> 上例中，`a` 的类型是 int，而 `a+1` 的列类型是 bigint，如果将生成列的类型定为 int，就不会发生替换。
->
-> 关于类型转换规则，可以参见[表达式求值的类型转换](/functions-and-operators/type-conversion-in-expression-evaluation.md)。
+> 若待替换的表达式类型和生成列类型都是字符类型，但两种类型长度不同时，仍可通过将系统变量 [`tidb_enable_unsafe_substitute`](/system-variables.md#tidb_enable_unsafe_substitute-从-v630-版本开始引入) 设置为 `ON` 来允许其替换。配置该系统变量时，需要保证生成列计算得到的值严格满足生成列的定义，否则，可能因为长度不同，导致数据截断得到错误的结果。详情见 [issue#35490](https://github.com/pingcap/tidb/issues/35490#issuecomment-1211658886)。
```
Comment (TomShawn) [inline generated-columns.md:155]:
```suggestion
> 若待替换的表达式类型和生成列类型都是字符类型，但两种类型长度不同时，仍可通过将系统变量 [`tidb_enable_unsafe_substitute`](/system-variables.md#tidb_enable_unsafe_substitute-从-v630-版本开始引入) 设置为 `ON` 来允许其替换。配置该系统变量时，需要保证生成列计算得到的值严格满足生成列的定义，否则，可能因为长度不同，导致数据截断得到错误的结果。详情见 GitHub issue [#35490](https://github.com/pingcap/tidb/issues/35490#issuecomment-1211658886)。
```

File: system-variables.md [hunk 7]
```diff
@@ -2410,3 +2410,10 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
 - 是否持久化到集群：是
 - 默认值：`ON`
 - 这个变量用于控制计算窗口函数时是否采用高精度模式。
+
+### `tidb_enable_unsafe_substitute` <span class="version-mark">从 v6.3.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 默认值：`OFF`
```
Comment (TomShawn) [inline system-variables.md:2418]:
```suggestion
- 类型：布尔型
- 默认值：`OFF`
```

**Review by TomShawn — APPROVED** [review]
minor comments
rest LGTM

**Pull request merged**